### ISSUE TITLE
Template length doesnt fit cleanly into flowset

### ISSUE THREAD

**Issue by msheiny** [logstash-plugins/logstash-codec-netflow#6]
I haven't really dug into this but on our Cisco I saw issues with the following lines and ended up commenting them out:

>           # Template shouldn't be longer than the record and there should
>           # be at most 3 padding bytes
>           if template.num_bytes > length or ! (length % template.num_bytes).between?(0, 3)
>           @logger.warn("Template length doesn't fit cleanly into flowset", :template_id => record.flowset_id, :template_length => template.num_bytes, :record_length => length)
>            next
>          end

After I commented it out, I've been getting netflow collection without issue. Can anyone illuminate a setting I can look at to work-around this without commenting? Perhaps we can provide a new flag to the codec option that allows us to simply skip this check?

Thanks in advance for any advice.

**Comment by wrigby** [comment]
@msheiny Do you happen to have the output of the log messages from when this happened? I'm curious to see which check failed - was it that the template length was longer than the flowset length, or that there were too many padding bytes?

I've spent a bit of time in the netflow codec code recently, and I'd be happy to dig deeper into this and possibly generate a PR for the elastic team.

Packet captures of the offending flow reports would be awesome, too.

**Comment by BMWfan** [comment]
Hello together,

i have talked with msheiny about that problem. 

I have the same problem.

Here you can see a Screenshot:
![logstash log](https://cloud.githubusercontent.com/assets/5577540/7230906/01f11a30-e772-11e4-86db-93869f86dfff.png)

Best Regards

Daniel

**Comment by wrigby** [comment]
Weird. I'd love to see a packet capture and your flow export configuration. I just read through a bunch of the source last night, and didn't see anything that jumps out at me.

**Comment by msheiny** [comment]
@wrigby  - sorry for the delay in getting back to you. I'll try to get you some packet dumps this week. Thanks for following up so quickly.

**Comment by wrigby** [comment]
@msheiny no problem! Full disclosure - I'm not a maintainer, just a guy that spent a lot of time in this code recently.

I've played around with tcpreplay a bit for testing this, so it should be pretty trivial to replay your data back into Logstash.

**Comment by bodgit** [comment]
Given that the log message is claiming the template is bigger than the record rather than faulty padding, this is either a grievous parsing bug somewhere or I wonder if it's #9 rearing it's head?

Do you have >1 device sending Netflow at Logstash?

**Comment by jorritfolmer** [comment]
I've encountered this too lately:

1. Netflow v9 flowsets from an options template (flowset_id == 1)
    The warning message is:  `Template length doesn't fit cleanly into flowset {:template_id=>259, :template_length=>8, :record_length=>12, :level=>:warn}`

    The template_length is correctly calculated, but the record length calculation doesn't take into account the extra "option scope" field that is present in these records. From what I can see in `netflow.rb`,  `length = record.flowset_length - 4` cannot be also applied to data records with an options template because of the extra 4 byte "option scope" field.

   A Wireshark capture of the offending option template from nprobe. Note the template_length in the warning message above is correctly leaving out the option scope.

    * Flowset Id: Options Template (v9) (1)
    * Flowset Length: 24
    * Options template
        - Template Id: 259
        - Option Scope Length: 4
        - Option Length: 8
        - **Field 1/1 [Scope]: System**
           - **Scope Type: System**
           - **Length: 4**
        - Field 1/2: TOTAL_FLOWS_EXP
           - Type: TOTAL_FLOWS_EXP (42)
           - Length: 4
       - Field 2/2: TOTAL_PKTS_EXP
            - Type: TOTAL_PKTS_EXP (41)
            - Length: 4

    Wireshark capture of the corresponding data record. Note the record_length in the warning above doesn't take into account the extra 4 byte ScopeSystem field )`length = record.flowset_length - 4` code in netflow.rb should be `length = record.flowset_length - 8` when encountering a data record from an options template.

    * Flowset Id: (Data) (259)
    * Flowset Length: 16
    * Flow 1
        * **ScopeSystem: 00000000**
        * FlowsExp: 1
        * PacketsExp: 0

2. Concurrently receiving v9 data flowsets from both softflowd and nprobe with differing field lengths.

    Individually netflow data from either softflowd or nprobe alone are processed correctly. 

    However when pointing the nprobe flows to the logstash collector, in addition to the softflowd flows, the nprobe flows are ignored with a `Template length doesn't fit cleanly into flowset {:template_id=>257, :template_length=>53, :record_length=>736, :level=>:warn}`

   The source of the length mismatch seems to be a difference in length of the `input_snmp` and `output_snmp` fields, that isn't properly retreived from the `@templates[key]` array.

    | field | length from softflowd template (id=1024) | length from by nprobe template (id=257)
    |-------|---------------------------------------|----------------------
    | input_snmp | 4 bytes | 2 bytes 
    | output_snmp | 4 bytes | 2 bytes

    It seems these lengths are correctly decoded from both data template flowsets and put into the respective`@templates[key]` array. Putting a few debugging statements in the template flowset processing section of netflow.rb confirms this, it shows lengths of 2 or 4 whenever the template records are processed. 
    
    However, when processing the data flowsets, only lenghts of 4 are are returned by accessing the `@templates[key]` array.  It seems the size declerations from softflowd templates have overwritten those from nprobe, even though they have different keys in the templates array.

**Comment by bodgit** [comment]
I wonder if I've inadvertently passed back a reference to the original netflow field definition rather than creating a fresh copy when parsing/building the templates?

Also, is there any overlap in the template ID's between softflowd and nprobe? If so I think you'll get a last-man-wins scenario if both also use the same source ID, (commonly it's just set to 0 on a lot of collectors).

Originally when this was a filter it had access to the source IP address that sent the original UDP/TCP packet which was also used to build the composite key for storing templates however as a codec it doesn't have that information so two devices sending the same template ID with the same source ID will overwrite each other as they refresh their templates. Obviously if you have >1 of the same device this shouldn't be a problem as the templates should be consistent but if you're mixing and matching collectors then this gets a lot more interesting.

**Comment by jorritfolmer** [comment]
As far as the source_id and template_id are concerned there is no overlap. According to wireshark:

| field | softflowd | nprobe |
|------|------|-----|
| SourceId | 0 | 157 |
| Flowset Id | 1024 | 257 |

**Comment by bodgit** [comment]
On this line:

https://github.com/logstash-plugins/logstash-codec-netflow/blob/master/lib/logstash/codecs/netflow.rb#L236

If you make the line (basically the `.clone`):

```ruby
field = @fields[type].clone
```

Does that fix the conflicting length problem?

**Comment by jorritfolmer** [comment]
Yes! That seems to fix it!

Before:

| decode_netflow9() stage |  nprobe input_snmp field | softflowd input_snmp field |
|--------------------|------------------|-------|
| "when 0" |  2 bytes  | 4 bytes |
| "when 256..65535 |  2 bytes | 2 bytes (mismatch)|

After:

| decode_netflow9() stage |  nprobe input_snmp field | softflowd input_snmp field |
|--------------------|------------------|-------|
| "when 0" |  2 bytes  | 4 bytes |
| "when 256..65535 |  2 bytes | 4 bytes |

**Comment by jorritfolmer** [comment]
Ah you mean your PR that is still open since May...
I'll do some testing on it. Nprobe also supports IPFIX.

**Comment by bodgit** [comment]
I don't think I've made the same fix to the Netflow path in the code so I'll whip up a one line PR for that.

**Comment by bodgit** [comment]
@jorritfolmer For your first observation, I think I see the problem.

Basically when I build any template based on an options template I seem to be currently ignoring any scope fields in the template and only looking at the option fields. This explains why for your example the record length is 12 but the template length is only 8, there should be another 4 bytes read by the template at the beginning to parse that scope field, that way the sizes would match.

I'm guessing if a collector sends a packet with lots of interleaved templates and data flowsets then the presence of an options template probably throws off parsing the rest of the packet depending where in the packet it is.

I must confess I wasn't entirely sure what to do with these options flows. I'll have to re-examine the spec and see how they're meant to be used.

**Comment by bodgit** [comment]
@jorritfolmer I've made some quick additional changes based on the clone feature branch to add some basic scope field parsing here:

https://github.com/bodgit/logstash-codec-netflow/tree/scope

You should get an addition `scope_foo` field added with an unsigned integer value in your option flowsets. Could you see if that fixes the problem with the option templates?

**Comment by romale** [comment]
@bodgit i've applied your suggestions: this https://github.com/logstash-plugins/logstash-codec-netflow/issues/6#issuecomment-132188679 and this https://github.com/logstash-plugins/logstash-codec-netflow/issues/6#issuecomment-133426176 , but no result. After code changing, i don't see error messages in logstash.log, but there is no netflow data in elastic from softflowd

**Comment by romale** [comment]
p.s. from cisco router i've netflow data in this instance

**Comment by jorritfolmer** [comment]
@bodgit Yes that look good! Using nprobe with v9 netflow, the JSON output from Logstash looks something like this:

````
{
  "@timestamp":"2015-08-22T10:50:39.000Z",
  "netflow": 
  {
    "version":9,
    "flow_seq_num":228,
    "flowset_id":259,
    "scope_system":0,
    "total_flows_exp":3010,
   "total_pkts_exp":228
  },
  "@version":"1",
  "host":"52.16.yy.zz"
}
````

Regarding the scope_foo fieldname: I think it would be better to rename the `scope_foo` field to `scope_type: foo`. Since the unsigned int already represents the scope type, there is probably no need to duplicate it in the fieldname. It also makes for easier grouping on this field if it would be:
 `"scope_type": "system"` or `"scope_type": 0" `

**Comment by bodgit** [comment]
This is the (brief) documentation I based that change on:

> Scope Field 1 Type
This field gives the relevant portion of the NetFlow process to which the options record refers. Currently defined values follow:
• 0x0001 System
• 0x0002 Interface
• 0x0003 Line Card
• 0x0004 NetFlow Cache
• 0x0005 Template
For example, sampled NetFlow can be implemented on a per-interface basis, so if the options record was reporting on how sampling is configured, the scope for the report would be 0x0002 (interface).

(Taken from http://www.cisco.com/en/US/technologies/tk648/tk362/technologies_white_paper09186a00800a3db9.html)

Based on that I don't think the field type is the same as the value as there's no `0` field type. As I understood it, you could push an option template indicating the scope field type was 0x0002/interface, and then the value of that field in the option record would indicate the interface index that would possibly/probably match the same value that is used for the `INPUT/OUTPUT_SNMP` field value in regular data records.

In your example I think the `0` you're getting for `scope_system` is the same value as that of the source ID in the header which for most software-based collectors seems to always be zero. Some example netflow packets from something with a non-zero source ID would confirm that.

**Comment by jorritfolmer** [comment]
Yes please ignore my previous post, I see what you mean, thanks for clarifying.

**Comment by jorritfolmer** [comment]
I've been doing some more testing, and get a:

```
Exception in inputworker {"exception"=>#<NoMethodError: undefined method `include?' for nil:NilClass>,
 "backtrace"=>["/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:442:in `netflow_field_for'",
"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:194:in `decode_netflow9'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/bindata-2.1.0/lib/bindata/array.rb:208:in `each'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/bindata-2.1.0/lib/bindata/array.rb:208:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:193:in `decode_netflow9'", "org/jruby/RubyKernel.java:1274:in `catch'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:191:in `decode_netflow9'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/bindata-2.1.0/lib/bindata/array.rb:208:in `each'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/bindata-2.1.0/lib/bindata/array.rb:208:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:190:in `decode_netflow9'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:133:in `decode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/bindata-2.1.0/lib/bindata/array.rb:208:in `each'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/bindata-2.1.0/lib/bindata/array.rb:208:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-netflow-1.0.0/lib/logstash/codecs/netflow.rb:132:in `decode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-udp-1.0.0/lib/logstash/inputs/udp.rb:95:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-udp-1.0.0/lib/logstash/inputs/udp.rb:74:in `udp_listener'"], :level=>:error}
```

on an initial clean run of nprobe (-V9) with this being the offending packet:

```
Frame 1: 258 bytes on wire (2064 bits), 258 bytes captured (2064 bits)
Ethernet II, Src: Vmware_70:86:09 (00:0c:29:70:86:09), Dst: Vmware_b9:5c:8f (00:0c:29:b9:5c:8f)
Internet Protocol Version 4, Src: 172.16.32.201 (172.16.32.201), Dst: 172.16.32.202 (172.16.32.202)
User Datagram Protocol, Src Port: 51826 (51826), Dst Port: iop (2055)
    Source Port: 51826 (51826)
    Destination Port: iop (2055)
    Length: 224
    Checksum: 0xfd2c [correct]
    [Stream index: 0]
Cisco NetFlow/IPFIX
    Version: 9
    Count: 3
    SysUptime: 35.035384000 seconds
    Timestamp: Aug 30, 2015 09:08:41.000000000 CEST
    FlowSequence: 0
    SourceId: 213
    FlowSet 1
        FlowSet Id: Data Template (V9) (0)
        FlowSet Length: 156
        Template (Id = 257, Count = 18)
            Template Id: 257
            Field Count: 18
            Field (1/18): BYTES
            Field (2/18): PKTS
            Field (3/18): PROTOCOL
            Field (4/18): IP_TOS
            Field (5/18): TCP_FLAGS
            Field (6/18): L4_SRC_PORT
            Field (7/18): IP_SRC_ADDR
            Field (8/18): SRC_MASK
            Field (9/18): INPUT_SNMP
            Field (10/18): L4_DST_PORT
            Field (11/18): IP_DST_ADDR
            Field (12/18): DST_MASK
            Field (13/18): OUTPUT_SNMP
            Field (14/18): IP_NEXT_HOP
            Field (15/18): SRC_AS
            Field (16/18): DST_AS
            Field (17/18): LAST_SWITCHED
            Field (18/18): FIRST_SWITCHED
        Template (Id = 258, Count = 18)
            Template Id: 258
            Field Count: 18
            Field (1/18): BYTES
            Field (2/18): PKTS
            Field (3/18): PROTOCOL
            Field (4/18): IP_TOS
            Field (5/18): TCP_FLAGS
            Field (6/18): L4_SRC_PORT
            Field (7/18): INPUT_SNMP
            Field (8/18): L4_DST_PORT
            Field (9/18): OUTPUT_SNMP
            Field (10/18): SRC_AS
            Field (11/18): DST_AS
            Field (12/18): LAST_SWITCHED
            Field (13/18): FIRST_SWITCHED
            Field (14/18): IPV6_SRC_ADDR
            Field (15/18): IPV6_DST_ADDR
            Field (16/18): IPV6_SRC_MASK
            Field (17/18): IPV6_DST_MASK
            Field (18/18): IPV6_NEXT_HOP
    FlowSet 2
        FlowSet Id: Options Template(V9) (1)
        FlowSet Length: 24
        Options Template (Id = 259) (Scope Count = 1; Data Count = 2)
            Template Id: 259
            Option Scope Length: 4
            Option Length: 8
            Field (1/1) [Scope]: System
            Field (1/2): TOTAL_FLOWS_EXP
            Field (2/2): TOTAL_PKTS_EXP
        [Padding]
    FlowSet 3
        FlowSet Id: (Data) (259)
        FlowSet Length: 16
        Flow 1
            ScopeSystem: 00000000
            FlowsExp: 1
            PacketsExp: 0
```

The .pcap for replay or wireshark:

```
cat << EOF |base64 -d |gunzip -dc > packet_NoMethodError_include_1pkt.pcap
H4sICKKx4lUAA2Z1enpfTm9NZXRob2RFcnJvcl9pbmNsdWRlLnBjYXAAfZAtDsIwHMVft451MGAjgZ
DwEewIB0CRmR1hjkMQDoHAcAEE6KFAoVFMgyc7A2aCjFfCDIImv/b9X9vXj9vluDcgUbaiAATHPMni
7GTBYKGBG5znG/aLlaMQAU8gROjvooM3upI0XSobj9cEDkxgPY2zJP+G3slWCLQYLaGPk1QWqRCbjq
LjUFepa8RlXWfdoG5Se8QnbdKBFMafrDJD7/3Zhx6dPhlw7ZDMqAW6wuSc+iSOSaA/hJ73vb9+P97l
mZxXKgEAAA==
EOF
```

**Comment by bodgit** [comment]
The only line that could be is:

```ruby
  def netflow_field_for(type, length)
    if @netflow_fields.include?(type) <-
      field = @netflow_fields[type]
      if field.is_a?(Array)
```

For `@netflow_fields` to be `nil` something weird must've happened as it should contain at least the default definitions.

**Comment by jorritfolmer** [comment]
Apologies! That weirdness is me manually mismerging the conflicts between clone and ipfix when merging locally. Sigh...

**Comment by hkshirish** [comment]
Can logstash netflow codec able to  handle netflow from different sources?

**Comment by bodgit** [comment]
Unless they all emit exactly the same template formats for any given template ID then you will need to be able to vary the source ID in the header. So if your different sources are all entirely the same piece of hardware or software then in theory it shouldn't be a problem.

The templates *should* be keyed on a tuple of `<source IP>,<source ID>,<template ID>` however the codec does not have access to the source IP address of the packet so they are currently keyed on `<source ID>,<template ID>` only.

Unfortunately a lot of exporters don't allow you to set the source ID and it's more often than not set to zero. However some exporters use template ID's 256 for options, 1024 for IPv4 flows and 2048 for IPv6 flows whereas some use 256, 257, 258, etc. so you may be lucky with what you've got.

**Comment by optiz0r** [comment]
In case this is useful to anyone else who comes across this issue, I was seeing this error message too, but in my case it was caused by some of the data fields being sent my my netflow device (mikrotik router) being different sizes than the netflow.yaml definitions file shipped with the netflow plugin.

For reference, I made the following changes:
```
# Changed from 2->4
10:
- 4
- :input_snmp
14:
- 4
- :output_snmp

# Changed from uint24->uint32
31:
- :uint32
- :ipv6_flow_label
```

I also added these entries:
```
225:
- :ip4_addr
- :ipv4_src_addr_postnat
226:
- :ip4_addr
- :ipv4_dst_addr_postnat
227:
- :uint16
- :l4_src_port_postnat
228:
- :uint16
- :l4_dst_port_postnat
```

**Comment by ssi0202** [comment]
hi is there a newer version available on githup that fixes this

**Comment by jorritfolmer** [comment]
There are patches waiting to be merged into a newer version. If you want a newer version, you'll have to patch and merge it yourself unfortunately.

**Comment by jorritfolmer** [comment]
This has been fixed now. In summary: there were multiple occurences of this issue:

* when receiving Netflow options templates with scope fields: fixed with PR #37 
* when receiving Netflow data from multiple devices with differing templates: fixed with PR #25
